### DISCUSSION TITLE
How to set vApp options in rbvmomi2

### DISCUSSION THREAD

**Discussion by bastelfreak** [ManageIQ/rbvmomi2#51] [Q&A]
Hi!

(I'm not a vmware expert, so my terminology might be wrong)

I maintain [beaker-vcloud](https://github.com/voxpupuli/beaker-vcloud) and [beaker-vmware](https://github.com/voxpupuli/beaker-vmware). Those tools basically create short-lived VMs where some integration tests are executed. we use the following to create a VM from a template:

https://github.com/voxpupuli/beaker-vcloud/blob/e9ab7fc8734066338414abe18df537e48f1ab36a/lib/beaker/hypervisor/vcloud.rb#L74-L163

Now I need to adjust it to pass vApp options as well. Is that supported by rbvmomi2 at the moment? I'm already passing guest information like this:

```ruby
extraConfig: [ { key: 'guestinfo.hostname', value: host['vmhostname'], },],
```
 (https://github.com/voxpupuli/beaker-vcloud/blob/e9ab7fc8734066338414abe18df537e48f1ab36a/lib/beaker/hypervisor/vcloud.rb#L82-L85)

But that's not exactly the same as vApp options? I can quest for the above attribute via `vmtoolsd --cmd "info-get guestinfo.hostname"`. But vApp Options are at the `guestinfo.ofvenv.*` namespace? And I cannot write them via `extraConfig: [ { key: 'guestinfo.ovfenv.hostname', value: host['vmhostname'], },],`?

**Comment by bastelfreak** [comment]
Edit: this is a similar request as https://github.com/vmware/govmomi/issues/2261, just for the ruby lib, not for go.

I just tried it again with:

```ruby
RbVmomi::VIM.VirtualMachineConfigSpec(
        annotation: 'Base template:  ' + host['template'] + "\n" +
          'Creation time:  ' + Time.now.strftime('%Y-%m-%d %H:%M') + "\n\n" +
          'CI build link:  ' + (ENV['BUILD_URL'] || 'Deployed independently of CI') +
          'department:     ' + @options[:department] +
          'project:        ' + @options[:project],
        extraConfig: [
          {
            key: 'guestinfo.ovfenv.test1',
            value: 'foo',
          },
          {
            key: 'guestinfo.test2',
            value: 'bar',
          },
        ],
      )
```

And `vmtoolsd --cmd "info-get guestinfo.test2"` returns `bar`, `vmtoolsd --cmd "info-get guestinfo.ovfenv"` returns the default vApp options from the template. There's an Option `test1`, but not set to `foo`. Interestingly `vmtoolsd --cmd "info-get guestinfo.ovfenv.test1"` works and returns `foo`.

**Reply by agrare** [reply]
@bastelfreak just to make sure I'm following, you want to set vApp Options in the same way that you would through the vSphere Web Client here?
![image](https://github.com/ManageIQ/rbvmomi2/assets/12851112/55c9de04-8ad7-4d9b-9f2e-ca01fa70a0ae)

If so those aren't set in extraConfig there is a `config.vAppConfig` property that backs that.

If you check the VMware SDK docs, the `config` attribute is a [`VirtualMachineConfigInfo`](https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U2/html/SDK/vsphere-ws/docs/ReferenceGuide/vim.vm.ConfigInfo.html), which has a `vAppConfig` property of type [`VmConfigInfo`](https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U2/html/SDK/vsphere-ws/docs/ReferenceGuide/vim.vApp.VmConfigInfo.html)

![image](https://github.com/ManageIQ/rbvmomi2/assets/12851112/1eb64b94-dcb2-4f4f-ac9a-401a9e4ecf92)

As an example I addes some product info:
![image](https://github.com/ManageIQ/rbvmomi2/assets/12851112/4d742707-04ef-4df6-896c-93205bebd802)

And some properties:
![image](https://github.com/ManageIQ/rbvmomi2/assets/12851112/a5ba21c1-9d65-4355-9d01-ae9ef16ebef6)

via the web UI so you can see what the underlying [`VAppProductInfo`](https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U2/html/SDK/vsphere-ws/docs/ReferenceGuide/vim.vApp.ProductInfo.html) and [`VAppPropertyInfo`](https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U2/html/SDK/vsphere-ws/docs/ReferenceGuide/vim.vApp.PropertyInfo.html) look like.

If you are on your vCenter you can use the ManagedObjectBrowser (MOB) to view what the API looks like, I find this immensely helpful, for example this vAppConfigInfo is at `https://VCENTER/mob/?moid=vm-1255&doPath=config.vAppConfig`

From here it is a "simple" [`reconfigureVM_Task`](https://dp-downloads.broadcom.com/api-content/apis/API_VWSA_001/8.0U2/html/SDK/vsphere-ws/docs/ReferenceGuide/vim.VirtualMachine.html#reconfigure) to change it.  I'll work up a little ruby example but if you're already doing reconfigures which it looks like you are it is just setting another attribute.

**Reply by bastelfreak** [reply]
> just to make sure I'm following, you want to set vApp Options in the same way that you would through the vSphere Web Client here?

Yes exactly!

My latest approach was:

```ruby
      vAppConfigSpec = RbVmomi::VIM.VmConfigSpec(
        property: [RbVmomi::VIM.VAppPropertySpec(
          operation: 'add',
          info: RbVmomi::VIM.VAppPropertyInfo(
            key: 1,
            category: 'network',
            type: 'hostname',
            value: 'test',
          ),
        )],
      )

      configSpec = RbVmomi::VIM.VirtualMachineConfigSpec( vAppConfig: vAppConfigSpec)
```

But with that the VM isn't created at all (and my current user has no option to see the error or I do not find it in the Web UI)

**Reply by bastelfreak** [reply]
This fails with:
![2024-05-28_13-08](https://github.com/ManageIQ/rbvmomi2/assets/341866/ffde25ce-7e9f-4169-ab35-7fa3dabb16dc)

After clicking a bit around I created a VM by hand:

![2024-05-28_13-13](https://github.com/ManageIQ/rbvmomi2/assets/341866/fe96df55-dd73-4d5e-a660-b122c6d82bdd)

so It seems like the options are a bit nested or grouped, into`puppet` and `network`. maybe that's the `category` attribute :thinking: 

Edit: mhm no. I already set `category: 'network',` and that created above error.

**Reply by agrare** [reply]
So `type` has to be something like `string` or `int` per https://vdc-repo.vmware.com/vmwb-repository/dcr-public/d1902b0e-d479-46bf-8ac9-cee0e31e8ec0/07ce8dbd-db48-4261-9b8f-c6d3ad8ba472/vim.vApp.PropertyInfo.html that's why you got that `invalid type hostname for property` error

**Reply by bastelfreak** [reply]
I'm not yet sure about the differences between `classId` vs `id` vs `instanceId`. My latest approach was:

```ruby
      vAppConfigSpec = RbVmomi::VIM.VmConfigSpec(
        property: [RbVmomi::VIM.VAppPropertySpec(
          operation: 'add',
          info: RbVmomi::VIM.VAppPropertyInfo(
            key: 1,
            category: 'network',
            type: 'string',
            value: 'test',
            classId: 'hostname',
            id: 'hostname'
          ),
        )],
      )

      configSpec = RbVmomi::VIM.VirtualMachineConfigSpec(vAppConfig: vAppConfigSpec)
```

This at least creates the VM, but when I afterwards check the VM in the vsphere client the vApp option isn't set.

**Reply by bastelfreak** [reply]
from the MOB. I got to VirtualMachineConfigInfo, then vAppConfig:
![2024-05-28_18-27](https://github.com/ManageIQ/rbvmomi2/assets/341866/7d5357d0-f378-4caf-b226-c4e183adb540)
and second property is the hostname I want to set:
![2024-05-28_18-26](https://github.com/ManageIQ/rbvmomi2/assets/341866/d63669d7-70a1-4bc9-a00d-816260759cee)

**Reply by bastelfreak** [reply]
I'm not really sure if I need to wrap `RbVmomi::VIM.VAppPropertyInfo` in `RbVmomi::VIM.VAppPropertySpec`?

**Reply by bastelfreak** [reply]
@agrare do you've any hints here for me?

**Reply by bastelfreak** [reply]
@agrare @Fryguy hey hey, any chance one of you can take a look here? I'm a bit lost in those nested objects :(

**Comment by Fryguy** [comment]
@agrare Can you take a look here?

@bastelfreak I moved this discussion from the general ManageIQ discussions page to this one specifically for rbvmomi2 (you didn't miss it - I just created it :D )
